### PR TITLE
Feature hide bots

### PR DESCRIPTION
**Pull request by VincePaulin**

### CODE REVIEW

File: lib/pages/chat_list/chat_list.dart [hunk 1]
```diff
@@ -159,10 +161,26 @@ class ChatListController extends State<ChatList>
...
+    return userHostName;
+  }
+
+  List<String> getBotIds() {
+    final hostName = getHostName();
+    return SocialNetworkManager.socialNetworks.map((sn) => sn.chatBot + hostName).toList();
+  }
+
+  // List of user or bot IDs to exclude
+  List<String> get excludedUserIds => getBotIds();
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: Performance : is there a way to calculate this only once, instead of at every call ?

**Review by ignyx — CHANGES_REQUESTED** [review]
Good code quality !

In my case the bots are not all filtered out (WhatsApp is). You may want to handle cases where there are multiple conversations with the bot.
image

File: lib/pages/chat_list/chat_list.dart [hunk 2]
```diff
@@ -454,6 +522,9 @@ class ChatListController extends State<ChatList>
 
     _checkTorBrowser();
 
+    // Calculates the list of excluded IDs at once
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: `at once` means `tout de suite`, consider leaving `excluded IDs once`

File: lib/pages/chat_list/chat_list.dart [hunk 3]
```diff
@@ -170,18 +172,84 @@ class ChatListController extends State<ChatList>
...
+    final Map<String, List<Room>> roomMap = {};
+    final client = Matrix.of(context).client;
+
+    // Group rooms by the bot ID in _excludedUserIds
+    for (var room in rooms) {
+      String? botId = room.directChatMatrixID;
+      if (botId == null && isGroupWithOnlyBotAndUser(room)) {
+        botId = _excludedUserIds.firstWhere(
+              (id) => room.getParticipants().any((user) => user.id == id),
+          orElse: () => '',
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: code would be cleaner if `orElse` resolved to `null`

File: lib/pages/chat_list/chat_list.dart [hunk 4]
```diff
@@ -170,18 +172,84 @@ class ChatListController extends State<ChatList>
...
+        roomList.sort((a, b) => b.lastEvent?.originServerTs.compareTo(a.lastEvent!.originServerTs) ?? 0);
+
+        // Keep the room with the most recent event by the bot
+        for (int i = 1; i < roomList.length; i++) {
+          final room = roomList[i];
+          if (room.lastEvent?.senderId == client.userID) {
+            await room.leave();
+          } else {
+            await roomList[0].leave();
+            roomList.removeAt(0);
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: I don't understand what is going on here. Why are you removing the first room ?

File: lib/pages/chat_list/chat_list.dart [hunk 5]
```diff
@@ -170,18 +172,84 @@ class ChatListController extends State<ChatList>
...
+        .toList();
 
-  List<Room> get filteredRooms => Matrix.of(context)
-      .client
-      .rooms
-      .where(getRoomFilterByActiveFilter(activeFilter))
-      .where((room) => !excludedUserIds.contains(room.directChatMatrixID))
-      .toList();
+    // Identify and remove duplicates
+    identifyAndRemoveDuplicates(filteredRooms);
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: The `filteredRooms` getter is called often and slows down app. Consider calling `identifyAndRemoveDuplicates` somewhere else.

**Review by ignyx — CHANGES_REQUESTED** [review]

File: lib/pages/chat_list/chat_list.dart (hunk 4 above)
Comment (VincePaulin) [reply lib/pages/chat_list/chat_list.dart]: The if was a very bad way to go when I think back on it. I made sure to keep the oldest conversation and went through the loop again.

File: lib/pages/chat_list/chat_list.dart (hunk 3 above)
Comment (ignyx) [reply lib/pages/chat_list/chat_list.dart]: casting them all to a string defeats the purpose. My bad, the original was good :+1:

**Comment by ignyx** [comment]
When clicking the back button (top left corner) in the bot conversations, nothing happens. Pressing the system back button works as expected. 
image

File: lib/pages/add_bridge/bot_chat_list.dart [hunk 6]
```diff
@@ -1,35 +1,78 @@
...
+          } else {
+            final rooms = snapshot.data!;
+            return ListView.builder(
+              itemCount: rooms.length,
+              itemBuilder: (context, i) {
+                final room = rooms[i];
+                return ChatListItem(rooms[i],
+                    key: Key('chat_list_item_${rooms[i].id}'), onTap: () {
+                      // Handle tap on a bot conversation
+                      _onChatTap(room, context);
```
Comment (ignyx) [inline lib/pages/add_bridge/bot_chat_list.dart]: consider renaming `_onChatTap` to what it does

File: lib/pages/add_bridge/bot_chat_list.dart [hunk 7]
```diff
@@ -28,12 +28,15 @@ class _BotChatListPageState extends State<BotChatListPage> {
...
-            .toList(),
+            () =>
+            Matrix
+                .of(context)
+                .client
+                .rooms
+                .where(
+                    (room) =>
+                    widget.botUserIds.contains(room.directChatMatrixID))
+                .toList(),
```
Comment (ignyx) [inline lib/pages/add_bridge/bot_chat_list.dart]: note : you keep committing linting operations

File: lib/pages/add_bridge/bot_chat_list.dart [hunk 8]
```diff
@@ -122,19 +125,19 @@ class _BotChatListPageState extends State<BotChatListPage> {
...
       // Reload rooms after leaving one
       _loadRooms();
       scaffoldMessenger.showSnackBar(
         SnackBar(
-          content: Text('Vous avez quitté la conversation avec succès.'),
+          content: Text(L10n.of(context)!.leaveTheConvSuccess),
         ),
       );
     } catch (e) {
       scaffoldMessenger.showSnackBar(
```
Comment (ignyx) [inline lib/pages/add_bridge/bot_chat_list.dart:140]: log the error

File: lib/pages/chat_list/chat_list.dart [hunk 9]
```diff
@@ -193,10 +196,29 @@ class ChatListController extends State<ChatList>
 
   bool isGroupWithOnlyBotAndUser(Room room) {
     final client = Matrix.of(context).client;
+
+    // Check that participants are fully charged
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart:204]: The next few lines seem bad for performance, especially during initial sync. Are they really necessary ? If its just to fix a few edge cases I'd rather them be removed.

File: lib/pages/chat_list/chat_list.dart [hunk 10]
```diff
@@ -193,10 +196,29 @@ class ChatListController extends State<ChatList>
...
-        participants.any((user) => user.id == client.id.toString());
+
+    if (participants.length != 2) {
+      return false;
+    }
+
+    // Check whether participants include the current user and one of the bots
+    final userIds = participants.map((user) => user.id).toList();
+    final containsCurrentUser = userIds.contains(client.userID);
+    final containsBot = userIds.any((id) => excludedUserIds.contains(id));
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: minor optimization can be to check `id.contains('bot')` (roughly) before checking `excludingUserIds.contains(id)`, which is more intensive

File: lib/pages/chat_list/chat_list.dart [hunk 11]
```diff
@@ -193,10 +196,29 @@ class ChatListController extends State<ChatList>
...
+
+    return containsCurrentUser && containsBot;
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: good :+1:

**Review by ignyx — CHANGES_REQUESTED** [review]

File: lib/pages/chat_list/chat_list.dart [hunk 12]
```diff
@@ -159,12 +168,107 @@ class ChatListController extends State<ChatList>
...
+
+  bool isGroupWithOnlyBotAndUser(Room room) {
+    final client = Matrix.of(context).client;
+
+    // Check that participants are fully charged
+    if (!room.participantListComplete && !loadingRooms.contains(room.id)) {
+      loadingRooms.add(room.id);
+      room.requestParticipants().then((_) {
+        loadingRooms.remove(room.id);
+        setState(() {}); // Force a rebuild to recheck the room
```
Comment (ignyx) [inline lib/pages/chat_list/chat_list.dart]: `setState` is still potentially called a high number of times ! I thought we agreed to only rebuild once after all the Futures are resolved (which is also done)

**Review by ignyx — CHANGES_REQUESTED** [review]
Almost done ! (State rebuild + error log)

**Pull request merged**